go/mysql/replication_status.go

```diff
@@ -34,19 +34,43 @@ type ReplicationStatus struct {
 	FilePosition         Position
 	FileRelayLogPosition Position
 	MasterServerID       uint
-	IOThreadRunning      bool
 	SQLThreadRunning     bool
 	SecondsBehindMaster  uint
 	MasterHost           string
 	MasterPort           int
 	MasterConnectRetry   int
 	MasterUUID           SID
+	IOThreadRunningState IOThreadRunningState
 }
 
-// ReplicationRunning returns true iff both the IO and SQL threads are
+// IOThreadRunningState communicates one of the three states returned from SHOW SLAVE STATUS for IO_Thread_Running, of either
+// "Yes", "Connecting", or "No".
+type IOThreadRunningState string
+
+var (
+	// IOThreadRunning communicates that the IO thread is running.
+	IOThreadRunning IOThreadRunningState = "Yes"
+	// IOThreadNotRunning communicates that the IO thread is not running.
+	IOThreadNotRunning IOThreadRunningState = "No"
+	// IOThreadConnecting communicates that the IO thread is currently attempting to re-connect.
+	IOThreadConnecting IOThreadRunningState = "Connecting"
+)
+
+// ReplicationRunning returns true if both the IO and SQL threads are
 // running.
 func (s *ReplicationStatus) ReplicationRunning() bool {
-	return s.IOThreadRunning && s.SQLThreadRunning
+	return s.IOThreadRunning() && s.SQLThreadRunning
+}
+
+// IOThreadRunning returns true if IO_Thread_Running is either "Yes", or "Connecting".
+func (s *ReplicationStatus) IOThreadRunning() bool {
+	switch s.IOThreadRunningState {
+	case IOThreadNotRunning:
+		return false
+	default:
+		// Return true for either "Connecting" or "Yes"
+		return true
+	}
```

Comment:
I think this will cause us to again get incorrect conclusions about replication. If this function returns `true` for `Connecting`, and `ReplicationRunning()` above returns `s.IOThreadRunning() && s.SQLThreadRunning`, then we conclude that `Connecting` => replication running. Was that not the original situation this PR was meant to fix?

I think we should return `false` for `Connecting`, so as to make a stricter analysis of the status. This will probably break something else somewhere else - but I think we should go down that rabbit hole nonetheless.